File: python_xo/environment.py
import random
from typing import Literal

import gymnasium as gym

from python_xo.agent import Agent
from python_xo.owner import Owner


class NaughtsAndCrossesEnvironment:
    def __init__(
        self,
        board: list[list[Owner]],
        user_mark: Literal[Owner.CROSS, Owner.NAUGHT],
        agent: Agent,
        time_step: int = 0,
    ) -> None:
        self.board = board
        self.time_step = time_step
        self.user_mark = user_mark
        self.agent = agent
        self.last_player = agent.agent_mark

    def reset(self) -> None:
        self.board = [
            [Owner.EMPTY, Owner.EMPTY, Owner.EMPTY],
            [Owner.EMPTY, Owner.EMPTY, Owner.EMPTY],
            [Owner.EMPTY, Owner.EMPTY, Owner.EMPTY],
        ]
        self.time_step = 0

        # 50/50 change of the agent taking the first move
        if random.choice([True, False]):
            self.agent_take_turn(self.agent.take_action(self.observation()))

    def agent_take_turn(self, action: int):
        row = action // 3
        col = action % 3
        if self.board[row][col] == Owner.EMPTY:
            self.board[row][col] = self.agent.agent_mark

    def render(self) -> None:
        for row in self.board:
            for i, cell in enumerate(row):
                if cell == Owner.EMPTY:
                    print(" ", end="")
                elif cell == Owner.NAUGHT:
                    print("O", end="")
                elif cell == Owner.CROSS:
                    print("X", end="")

                if i < 2:
                    print("|", end="")

            print()

    def observation(self) -> tuple[int, int, int, int, int, int, int, int, int]:
        return (
            self.board[0][0].value,
            self.board[0][1].value,
            self.board[0][2].value,
            self.board[1][0].value,
            self.board[1][1].value,
            self.board[1][2].value,
            self.board[2][0].value,
            self.board[2][1].value,
            self.board[2][2].value,
        )

    def game_is_draw(self) -> bool:
        return all([cell != Owner.EMPTY for row in self.board for cell in row])

    def game_winner(self) -> Owner:
        for i in range(3):
            # Check for a win in the rows
            if (
                self.board[i][0] == self.board[i][1] == self.board[i][2]
                and self.board[i][0] != Owner.EMPTY
            ):
                return self.board[i][0]

            # Check for a win in the columns
            if (
                self.board[0][i] == self.board[1][i] == self.board[2][i]
                and self.board[0][i] != Owner.EMPTY
            ):
                return self.board[0][i]

        # Check for a win in the diagonals
        if (
            self.board[0][0] == self.board[1][1] == self.board[2][2]
            and self.board[0][0] != Owner.EMPTY
        ):
            return self.board[0][0]

        if (
            self.board[0][2] == self.board[1][1] == self.board[2][0]
            and self.board[0][2] != Owner.EMPTY
        ):
            return self.board[0][2]

        return Owner.EMPTY

    def terminated(self) -> bool:
        return self.game_is_draw() or self.game_winner() != Owner.EMPTY

    def truncated(self) -> bool:
        return self.time_step > 9

    def reward(self) -> int:
        match (self.game_winner(), self.game_is_draw()):
            case (Owner.NAUGHT, _):
                return -100
            case (Owner.CROSS, _):
                return 100
            case (_, True):
                return 50
            case _:
                return 0

    def step(
        self, action: int
    ) -> tuple[tuple[int, int, int, int, int, int, int, int, int], int, bool, bool]:
        if not self.terminated():
            row = action // 3
            col = action % 3
            if self.board[row][col] == Owner.EMPTY:
                self.board[row][col] = self.user_mark

            if not self.terminated():
                self.agent_take_turn(self.agent.take_action(self.observation()))

        self.time_step += 1

        return self.observation(), self.reward(), self.terminated(), self.truncated()


class NaughtsAndCrossesEnvironmentGym(gym.Env):
    def __init__(self, naughts_and_crosses_environment: NaughtsAndCrossesEnvironment) -> None:
        self.naughts_and_crosses_environment = naughts_and_crosses_environment
        self.action_space = gym.spaces.Discrete(9)
        self.observation_space = gym.spaces.MultiDiscrete([3] * 9)

    def reset(self, seed=None, options=None):
        super().reset(seed=seed)

        self.naughts_and_crosses_environment.reset()
        return self.naughts_and_crosses_environment.observation(), None

    def step(self, action: int):
        observation, reward, terminated, truncated = self.naughts_and_crosses_environment.step(
            action
        )
        return observation, reward, terminated, truncated, {}

    def render(self):
        self.naughts_and_crosses_environment.render()

    def close(self):
        pass

File: go_xo/environment.go
package xo

import (
	"fmt"
	"math/rand"
)

const (
	Empty = 0
	Naught = 1
	Cross = 2
)

type Board [3][3]int

type NaughtsAndCrossesEnvironment struct {
	Board
	timeStep int
}

func (b *NaughtsAndCrossesEnvironment) Reset() {
	b.Board = Board{
		{Empty, Empty, Empty},
		{Empty, Empty, Empty},
		{Empty, Empty, Empty},
	}
	b.timeStep = 0

	// Randomly decide if the agent goes first
	if rand.Intn(2) == 1 {
		row, col := b.randomSpace()
		b.Board[row][col] = Naught
	}
}

func (b *NaughtsAndCrossesEnvironment) randomSpace() (int, int) {
	row := rand.Intn(len(b.Board))
	col := rand.Intn(len(b.Board))
	return row, col
}

func (b *NaughtsAndCrossesEnvironment) agentTakeTurn() {
	// Select the first non-empty space
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if b.Board[i][j] == Empty {
				b.Board[i][j] = Naught
				return
			}
		}
	}
}

func (b *NaughtsAndCrossesEnvironment) Step(action int) (observation [9]int, reward int, terminated bool, truncated bool) {
    if !b.Terminated() {
        row := action / 3
        col := action % 3
        if b.Board[row][col] == Empty {
            b.Board[row][col] = Cross
        }
    }

    reward = b.Reward()
    terminated = b.Terminated()

    if !terminated {
        b.agentTakeTurn()
        b.timeStep += 1
    }

	observation = b.Observation()
	truncated = false

	return
}

func (b *NaughtsAndCrossesEnvironment) Render() {
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			switch b.Board[i][j] {
			case Empty:
				fmt.Print(" ")
			case Naught:
				fmt.Print("O")
			case Cross:
				fmt.Print("X")
			}

			if j < 2 {
				fmt.Print("|")
			}
		}
		fmt.Println()
	}
}

func (b *NaughtsAndCrossesEnvironment) Observation() [9]int {
	obs := [9]int{}

	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			obs[i*3+j] = b.Board[i][j]
		}
	}
	return obs
}

func (b *NaughtsAndCrossesEnvironment) gameIsDraw() bool {
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if b.Board[i][j] == Empty {
                return false
            }
		}
	}

	return true
}

func (b *NaughtsAndCrossesEnvironment) gameWinner() int {
	for i := 0; i < 3; i++ {
		// Check for a win in the rows
        if b.Board[i][0] == b.Board[i][1] && b.Board[i][1] == b.Board[i][2] && b.Board[i][0] != Empty{
            return b.Board[i][0]
        }

        // Check for a win in the columns
        if b.Board[0][i] == b.Board[1][i] && b.Board[1][i] == b.Board[2][i] && b.Board[0][i] != Empty{
            return b.Board[0][i]
        }
    }

    // Check for a win in the diagonals
    if b.Board[0][0] == b.Board[1][1] && b.Board[1][1] == b.Board[2][2] && b.Board[0][0] != Empty{
        return b.Board[0][0]
    }

    if b.Board[0][2] == b.Board[1][1] && b.Board[1][1] == b.Board[2][0] && b.Board[0][2] != Empty{
        return b.Board[0][2]
    }

	return Empty
}

func (b *NaughtsAndCrossesEnvironment) Terminated() bool {
	return b.gameIsDraw() || b.gameWinner() != Empty
}

func (b *NaughtsAndCrossesEnvironment) Reward() int {
	switch b.gameWinner() {
		case Naught:
			return -10
		case Cross:
			return 10
		}

	if b.gameIsDraw() {
		return 1
	}

	return 0
}